# Assignment 2

Student: Francisco Neves (up201404576)

## 1. Generative classifiers

Consider a classification problem with a target variable $y \in \{0, 1\}$ and input features $\boldsymbol{x} = (x_1\; x_2\; x_3\; x_4)^T$, where $x_1 \in \{0, 1\}$, $x_2 \in \{0, 1\}$, and $(x_3, x_4) \in \mathbb{R}^2$. Further assume that:
- $(x_1, x_2)$ is conditionally independent of $(x_3, x_4)$ given $y$;
- $x_1$ and $x_2$ are **dependent** given $y$;
- $x_3$ and $x_4$ are **dependent** given $y$;
- the conditional distributions of $(x_3, x_4)$ given $y$ are Gaussian.

### 1.1

**a)** Enumerate the parameters of the MAP classifier: $$\hat{y} = \text{arg} \max_{y \in \{0, 1\}} p(y)p(\boldsymbol{x} \mid y),$$ and indicate the dimension of each parameter.

___
Before estimating the parameters, by assuming partial independence between $(x_1,x_2)$ and $(x_3,x_4)$ the likelihood is described by:
$\begin{equation}
p\left(\begin{bmatrix}
x_{1} \\
x_{2} \\
x_{3} \\
x_{4}
\end{bmatrix}\middle\vert y\right) = 
p\left(\begin{bmatrix}
x_{1} \\
x_{2}
\end{bmatrix}\middle\vert y\right)
p\left(\begin{bmatrix}
x_{3} \\
x_{4}
\end{bmatrix}\middle\vert y\right).
\end{equation}$

Following, we enumerate the different parameters by breaking the problem in three parts (a), (b) and (c).
### (a) Parameters for $(x_1,x_2)$
For the jointed $(x_1, x_2)$ features, we have six parameters in total: 3 parameters for $p\left(\begin{bmatrix} x_{1} \\ x_{2}\end{bmatrix}\middle\vert y = 0 \right)$ and 3 parameters for $p\left(\begin{bmatrix} x_{1} \\ x_{2}\end{bmatrix}\middle\vert y = 1\right)$. Since $x_1 \in \{0,1\}$ and $x_2 \in \{0,1\}$ and they are dependent between each other given $y$, one needs to consider all the possible four joint configuration using 2 bits of information, such as $[0,0], [0,1], [1,0], [1,1]$. At least three parameters are enough to estimate for each $p\left(\begin{bmatrix} x_{1} \\ x_{2}\end{bmatrix}\middle\vert y_i\right)$, since a fourth is indirectly deduced from the others, such as,
$\begin{equation}
  p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 0\end{bmatrix}\middle\vert y_i\right) + p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 1\end{bmatrix}\middle\vert y_i \right) + p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 0\end{bmatrix}\middle\vert y_i \right) + p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 1\end{bmatrix}\middle\vert y_i \right) = 1,
\end{equation}$
where if $p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 1\end{bmatrix}\middle\vert y_i \right)$ is the one not considered, the computation is as follows:
$\begin{equation}
p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 1\end{bmatrix}\middle\vert y_i \right) = 1 - \left(p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 0\end{bmatrix}\middle\vert y_i\right) + p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 1\end{bmatrix}\middle\vert y_i \right) + p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 0\end{bmatrix}\middle\vert y_i \right)\right).
\end{equation}$ 
Hence, the six parameters are the following:
* 3 parameters for $p\left(\begin{bmatrix} x_{1} \\ x_{2}\end{bmatrix}\middle\vert y = 0 \right)$: 
  * $p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 0\end{bmatrix}\middle\vert y = 0 \right)$
  * $p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 1\end{bmatrix}\middle\vert y = 0 \right)$
  * $p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 0\end{bmatrix}\middle\vert y = 0 \right)$
* 3 parameters for $p\left(\begin{bmatrix} x_{1} \\ x_{2}\end{bmatrix}\middle\vert y = 1 \right)$:
  * $p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 0\end{bmatrix}\middle\vert y = 1 \right)$
  * $p\left(\begin{bmatrix} x_{1} = 0 \\ x_{2} = 1\end{bmatrix}\middle\vert y = 1 \right)$
  * $p\left(\begin{bmatrix} x_{1} = 1 \\ x_{2} = 0\end{bmatrix}\middle\vert y = 1 \right)$


### (b) Parameters for $(x_3, x_4)$
For the jointed $(x_3, x_4)$ features, we have ten parameters, as follows: 
* 4 means: $\mu_{x_3}=\begin{bmatrix}\mu_{x_3}^{y=0} \\ \mu_{x_3}^{y=1}\end{bmatrix}$ and $\mu_{x_4}=\begin{bmatrix}\mu_{x_4}^{y=0} \\ \mu_{x_4}^{y=1}\end{bmatrix}$
* 6 covariance elements: $\begin{bmatrix}
\sigma(x_3,x_3)^{2}_{y = 0} & \sigma(x_3,x_4)^{2}_{y = 0} \\
\cdots & \sigma(x_4,x_4)^{2}_{y = 0}
\end{bmatrix}$ and 
$ \begin{bmatrix}
\sigma(x_3,x_3)^{2}_{y = 1} & \sigma(x_3,x_4)^{2}_{y = 1} \\
\cdots & \sigma(x_4,x_4)^{2}_{y = 1}
\end{bmatrix}$, where $ \sigma(x_4,x_3)^{2}_{y = 1}$ and  $\sigma(x_4,x_3)^{2}_{y = 0}$ are omitted, since  $\sigma(x_3,x_4)^{2}_{y = 0} = \sigma(x_4,x_3)^{2}_{y = 0}$ and $\sigma(x_3,x_4)^{2}_{y = 1} = \sigma(x_4,x_3)^{2}_{y = 1}$ because the covariance matrix is symmetric.

### (c) Prior parameters
Finally, we have to estimate just one prior, e.g,: $p(y=0)$. Since $p(y=0) = 1 - p(y=1)$, knowing one of them deduces the other.

### Number of parameters
Hence, in total the model has 17 parameters, such as: 6 parameters from the part (a) plus 10 parameters from the part (b) plus 1 parameter from the part (c).
___

**b)** Given a dataset $\{(\boldsymbol{x}^{(i)}, y^{(i)})\}_{i=1}^n$, write the expressions for the maximum likelihood estimates of the parameters enumerated in the previous question.

*YOUR ANSWER HERE* (either typeset or a digitalization of your handwritten answer)

### 1.2

In [7]:
# Libraries
import numpy as np

Now, you will implement this classifier in Python. The classifier skeleton is provided below in the class `Classifier`. You may implement additional auxiliary methods that you find useful.

In [119]:
class Classifier:

    def fit(self, X, y):
        '''
        Inputs:
            X - np.array with shape (num_examples_train, 4)
            y - np.array with shape (num_examples_train,)
        '''
        # Class attributes
        self.__priors = {}
        self.__mus = {}
        self.__covs = {}

        # Useful variables
        Xy = np.concatenate((X,y.reshape(-1,1)), axis = 1)
        target = Xy[:,-1]
        N = Xy.shape[0]
        X_0, X_1 = Xy[target == 0], Xy[target == 1]
        x00_0 = X_0[(X_0[:,0] == 0) & (X_0[:,1] == 0)]
        x01_0 = X_0[(X_0[:,0] == 0) & (X_0[:,1] == 1)]
        x10_0 = X_0[(X_0[:,0] == 1) & (X_0[:,1] == 0)]
        x11_0 = X_0[(X_0[:,0] == 1) & (X_0[:,1] == 1)]
        x00_1 = X_1[(X_1[:,0] == 0) & (X_1[:,1] == 0)]
        x01_1 = X_1[(X_1[:,0] == 0) & (X_1[:,1] == 1)]
        x10_1 = X_1[(X_1[:,0] == 1) & (X_1[:,1] == 0)]
        x01_1 = X_1[(X_1[:,0] == 1) & (X_1[:,1] == 1)]
        x3_0, x3_1, x4_0, x4_1 = X_0[:,2].reshape(-1,1), X_1[:,2].reshape(-1,1), X_0[:,3].reshape(-1,1), X_1[:,3].reshape(-1,1)

        # (a) Conditional discrete probabilities for x1 and x2
        p00_0, p01_0, p10_0 = x00_0.shape[0] / X_0.shape[0], x01_0.shape[0] / X_0.shape[0], x10_0.shape[0] / X_0.shape[0]
        p11_0 = 1 - (p00_0 + p01_0 + p10_0)
        p00_1, p01_1, p10_1 = x00_1.shape[0] / X_1.shape[0], x01_1.shape[0] / X_1.shape[0], x10_1.shape[0] / X_1.shape[0]
        p11_1 = 1 - (p00_1 + p01_1 + p10_1)

        # (b) Mean and covariance parameters for x3 and x4
        self.__mus['x3'] = {0: np.mean(x3_0, axis = 0).item(), 1: np.mean(x3_1, axis = 0).item()}
        self.__mus['x4'] = {0: np.mean(x4_0, axis = 0).item(), 1: np.mean(x4_1, axis = 0).item()}
        self.__covs[0] = np.array([
            [np.sum(np.square(x3_0 - self.__mus['x3'][0])) / N, np.sum((x3_0 - self.__mus['x3'][0]) * (x4_0 - self.__mus['x4'][0])) / N],
            [np.sum((x4_0 - self.__mus['x4'][0]) * (x3_0 - self.__mus['x3'][0])) / N, np.sum(np.square(x4_0 - self.__mus['x4'][0])) / N]
            ])
        self.__covs[1] = np.array([
            [np.sum(np.square(x3_1 - self.__mus['x3'][1])) / N, np.sum((x3_1 - self.__mus['x3'][1]) * (x4_1 - self.__mus['x4'][1])) / N],
            [np.sum((x4_1 - self.__mus['x4'][1]) * (x3_1 - self.__mus['x3'][1])) / N, np.sum(np.square(x4_1 - self.__mus['x4'][1])) / N]
            ])

        # (c) Prior parameter
        self.__priors[0] = len(y[y == 0]) / y.shape[0]
        self.__priors[1] = 1 - self.__priors[0]

        pass  # remove this line
    def bivariate_gaussian(x1, x2, mean1, mean2, cov_matrix):
        """
        std_dev_1 = np.sqrt(cov_matrix[0,0])
        std_dev_2 = np.sqrt(cov_matrix[1,1])
        rho = cov_matrix[0,1] / (mean1 * mean2)
        k = 1 / (2 * np.pi * mean1 * mean2 * np.sqrt(1 - rho ** 2))
        z = () - (2 * rho(x1 - mean1)) / (mean1 * mean2) + (x2 - mean2)**2 / (mean2 ** 2)
        np.exp()
        """
    
    def predict(self, X):
        '''
        Inputs:
            X - np.array with shape (num_examples_test, 4)
        
        Outputs:
            ypred - np.array with shape (num_examples_test,)
            posteriors - np.array with shape (num_examples_test, 2)
        '''
        #posteriors = []
        
        pass  # remove this line

**N.B.:** In both a) and b), you should avoid for loops as much as possible by using vectorized NumPy operations and broadcasting.

**a)** Implement the `fit` method, which receives as input two `np.array`s:
- `X`, which contains the 4-dimensional training input examples $\boldsymbol{x}^{(i)}$, one per row;
- `y`, which contains the corresponding training labels $y^{(i)} \in \{0,1\}$, one per row.

This method should compute the maximum likelihood estimates of the model parameters and store them as class attributes.

**b)** Implement the `predict` method, which receives as input one `np.array`:
- `X`, which contains the 4-dimensional examples $\boldsymbol{x}^{(i)}$ to be classified, one per row.

This function should return two `np.array`s:
- `ypred`, which should contain the labels predicted for each $\boldsymbol{x}^{(i)}$, one per row.
- `posteriors`, which should contain the posterior probabilities of each class given each $\boldsymbol{x}^{(i)}$, one per row.

If you have solved a) and b) correctly, the code below should run without errors and the reported test accuracy should be higher than 80%.

In [120]:
# read the data from file
data = np.genfromtxt('ex1_data.txt')
X, y = data[:, 0:4], data[:, 4].astype(int)

# use the first 400 lines for training and the remaining 100 lines for testing
Xtrain, ytrain = X[0:400], y[0:400]
Xtest, ytest = X[400:], y[400:]

# instantiate the classifier and train it
classifier = Classifier()
classifier.fit(Xtrain, ytrain)

# get the predictions on the test data
ypred, posteriors = classifier.predict(Xtest)
print('Example 0:')
print('  posteriors =', posteriors[0])
print('  predicted class =', ypred[0])
print('  ground-truth class =', ytest[0])
print()
print('Example 1:')
print('  posteriors =', posteriors[1])
print('  predicted class =', ypred[1])
print('  ground-truth class =', ytest[1])
print()
print('Example 2:')
print('  posteriors =', posteriors[2])
print('  predicted class =', ypred[2])
print('  ground-truth class =', ytest[2])
print()

# compute the accuracy on the test set
acc = np.mean(ypred == ytest)
print(f'Test accuracy = {100.*acc:.1f}%')

0.1072961373390558 0.24034334763948498 0.44635193133047213 0.20600858369098707
0.2994011976047904 0.10778443113772455 0.3652694610778443 0.22754491017964074


TypeError: cannot unpack non-iterable NoneType object

## 2. Logistic regression

Consider the `heightWeightData.txt` dataset that you have used in the Lab classes. You will use this data to build a Logistic Regression classifier that predicts the sex of an individual given their height and weight.

**a)** Train a Logistic Regression classifier **using only the first 160 rows** of the dataset as training data. You may use Scikit-Learn (`sklearn.linear_model.LogisticRegression`). **Print the values of the learned parameters.**

In [ ]:
from sklearn.linear_model import LogisticRegression
import pandas as pd

data = pd.read_csv('heightWeightData.txt', sep = ",", header = None)
X_train, Y_train, X_test, Y_test = data.iloc[:160,1:], data.iloc[:160,0], data.iloc[160:,1:], data.iloc[160:,0]

# 1. Creating the model
model = LogisticRegression(random_state = 0) 

# 2. Training the model
clf = model.fit(X_train, Y_train) 

# Printing the coefficients
print(clf.coef_, clf.intercept_)

[[-0.15294769 -0.10445359]] [34.00288831]


**b)** Compute the predictions of your model in the remaining 50 rows of the dataset and report the classification accuracy of your model in this test set.

In [ ]:
preds = clf.predict(X_test)
print(f'Predictions for the last 50 rows: \n{preds}')
print(f'The classification accuracy is {100 * model.score(X_test, Y_test)}%')

Predictions for the last 50 rows: 
[1 2 1 2 2 2 2 1 2 1 2 2 2 1 2 2 2 2 2 2 1 1 2 2 1 2 2 2 1 2 2 1 2 2 2 2 1
 2 2 1 2 2 1 2 1 1 1 1 2 1]
The classification accuracy is 92.0%


**c)** Using the parameter values printed in a), write the equation of the decision boundary of your model.

$Y = W^TX + W_0$,
where 
$W^T=[-0.15294769, -0.10445359]$ 
and $W_0 = 34.00288831$.